### PR TITLE
[red-knot] Report invalid exceptions

### LINKED ISSUE #15038: [red-knot] Add a diagnostic for `raise` statements used with non-exceptions
Red-knot currently doesn't emit a diagnostic on this snippet:

```py
raise 42
```

But we should, since `42` is not a valid object to be used in a `raise` statement:

```pycon
>>> raise 42
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    raise 42
TypeError: exceptions must derive from BaseException
```

The rule is that the object used in a `raise` statement must be assignable to the type `BaseException | type[BaseException]`, since both of the following work:

```pycon
>>> raise ValueError
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    raise ValueError
ValueError
>>> raise ValueError()
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    raise ValueError()
ValueError
```

We need to add some extra logic to this branch here: https://github.com/astral-sh/ruff/blob/c9fdb1f5e3a4d0d60b4537741f2c9c19e2426038/crates/red_knot_python_semantic/src/types/infer.rs#L2193-L2201

I think this will probably need to be a new rule added to https://github.com/astral-sh/ruff/blob/main/crates/red_knot_python_semantic/src/types/diagnostic.rs ? I don't think it fits neatly into any other rule. The alternative might be to make the existing `INVALID_EXCEPTION_CAUGHT` rule broader (and rename it).

### PR DESCRIPTION
**Pull request by InSyncWithFoo**
## Summary

Resolves #15038.

## Test Plan

Markdown tests.

### CODE REVIEW

File: crates/red_knot_python_semantic/resources/mdtest/exception/basic.md [hunk 1]
```diff
@@ -80,13 +80,72 @@ def foo(
...
+    raise SystemExit from bool()  # error: [invalid-exception]
+except:
+    ...
+
+try:
+    raise
+except int as e:  # error: [invalid-exception]
+    reveal_type(e)  # revealed: Unknown
+    raise LookupError from e
+```
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/exception/basic.md:151]: Might be nice to have a happy-path test similar to this but showing that we can catch an exception and `raise from` that exception without error. So like this test, but with a valid exception type in place of `int`.

File: crates/red_knot_python_semantic/src/types/diagnostic.rs [hunk 2]
```diff
@@ -199,35 +199,38 @@ declare_lint! {
 }
 
 declare_lint! {
-    /// Checks for exception handlers that catch non-exception classes.
+    /// Checks for `raise` statements and handlers
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/diagnostic.rs:202]:
```suggestion
    /// Checks for `raise` statements and exception handlers
```

File: crates/red_knot_python_semantic/src/types/diagnostic.rs [hunk 3]
```diff
@@ -720,9 +723,31 @@ impl<'db> TypeCheckDiagnosticsBuilder<'db> {
         );
     }
 
+    pub(super) fn add_invalid_exception_raised(&mut self, db: &dyn Db, node: &ast::Expr, ty: Type) {
+        self.add_lint(
+            &INVALID_EXCEPTION,
+            node.into(),
+            format_args!(
+                "Cannot raise object of type `{}` (must be a `BaseException` subclass)",
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/diagnostic.rs:731]:
```suggestion
                "Cannot raise object of type `{}` (must be a `BaseException` subclass or instance)",
```

File: crates/red_knot_python_semantic/src/types/diagnostic.rs [hunk 4]
```diff
@@ -720,9 +723,31 @@ impl<'db> TypeCheckDiagnosticsBuilder<'db> {
...
+            ),
+        );
+    }
+
+    pub(super) fn add_invalid_exception_cause(&mut self, db: &dyn Db, node: &ast::Expr, ty: Type) {
+        self.add_lint(
+            &INVALID_EXCEPTION,
+            node.into(),
+            format_args!(
+                "Cannot use object of type `{}` as exception cause (must be a `BaseException` subclass)",
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/diagnostic.rs:742]:
```suggestion
                "Cannot use object of type `{}` as exception cause (must be a `BaseException` subclass or instance, or `None`)",
```

**Review by carljm — COMMENTED** [review]
Nice! This looks good to me.

I'd like to get @AlexWaygood review on it, since he's done most of the work on exceptions so far. Feel free to hold off on implementing my comments until he reviews also.

File: crates/red_knot_python_semantic/src/types/diagnostic.rs [hunk 5]
```diff
@@ -720,9 +723,32 @@ impl<'db> TypeCheckDiagnosticsBuilder<'db> {
...
+            &INVALID_EXCEPTION,
+            node.into(),
+            format_args!(
+                "Cannot raise object of type `{}` (must be a `BaseException` subclass or instance)",
+                ty.display(db)
+            ),
+        );
+    }
+
+    pub(super) fn add_invalid_exception_cause(&mut self, db: &dyn Db, node: &ast::Expr, ty: Type) {
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types/diagnostic.rs:737]:
```suggestion
    pub(super) fn add_invalid_exception_caught(&mut self, db: &dyn Db, node: &ast::Expr, ty: Type) {
```

File: crates/red_knot_python_semantic/src/types/diagnostic.rs [hunk 6]
```diff
@@ -199,35 +199,38 @@ declare_lint! {
...
-    /// - [Python documentation: except clause](https://docs.python.org/3/reference/compound_stmts.html#except-clause)
+    /// - [Python documentation: The `raise` statement](https://docs.python.org/3/reference/simple_stmts.html#raise)
+    /// - [Python documentation: `except` clause](https://docs.python.org/3/reference/compound_stmts.html#except-clause)
     /// - [Python documentation: Built-in Exceptions](https://docs.python.org/3/library/exceptions.html#built-in-exceptions)
     ///
     /// ## Ruff rule
     ///  This rule corresponds to Ruff's [`except-with-non-exception-classes` (`B030`)](https://docs.astral.sh/ruff/rules/except-with-non-exception-classes)
-    pub(crate) static INVALID_EXCEPTION_CAUGHT = {
-        summary: "detects exception handlers that catch classes that do not inherit from `BaseException`",
+    pub(crate) static INVALID_EXCEPTION = {
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types/diagnostic.rs:232]: Seeing the changes to the documentation, I now think that these should be two error codes: `invalid-exception-caught` and `invalid-exception-raised`

File: crates/red_knot_python_semantic/src/types.rs [hunk 7]
```diff
@@ -1443,6 +1443,36 @@ impl<'db> Type<'db> {
...
+        self.is_assignable_to(db, subclass_of_base_exception)
+    }
+
+    pub(crate) fn is_exception(self, db: &'db dyn Db) -> bool {
+        let instance_of_base_exception = KnownClass::BaseException.to_instance(db);
+
+        self.is_assignable_to(db, instance_of_base_exception)
+    }
+
+    pub(crate) fn is_raisable(self, db: &'db dyn Db) -> bool {
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types.rs:1468]:
```suggestion
    pub(crate) fn can_be_raised(self, db: &'db dyn Db) -> bool {
```

File: crates/red_knot_python_semantic/src/types.rs [hunk 8]
```diff
@@ -1443,6 +1443,36 @@ impl<'db> Type<'db> {
...
+        self.is_exception(db) || self.is_exception_class(db)
+    }
+
+    pub(crate) fn can_be_exception_cause(self, db: &'db dyn Db) -> bool {
+        self.is_raisable(db) || self.is_none(db)
+    }
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types.rs:1474]: I'm interested in hearing other opinions. These methods feel too specific to me to have them on `Type`. That's why I'm leaning towards making them standalone functions. Unless we forsee using them in other places too (I want to avoid that `Type` ends up with a trillion very specific methods)

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 9]
```diff
@@ -2189,13 +2184,25 @@ impl<'db> TypeInferenceBuilder<'db> {
...
-        let ast::StmtRaise {
-            range: _,
-            exc,
-            cause,
-        } = raise;
-        self.infer_optional_expression(exc.as_deref());
-        self.infer_optional_expression(cause.as_deref());
+        let Some(raised) = &raise.exc else {
+            return;
+        };
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:2189]: As you've seen before in your PRs, red-knot panics if we fail to infer the types for all sub-nodes in an AST. The way you've written this method means that if `exc` is `None` but `cause` is `Some()`, red-knot will not infer the type of the `cause` sub-node, and might panic. Now, with valid Python syntax, it's impossible for `cause` to be `Some()` if `exc` is `None` -- but red-knot needs to be able to analyze invalid Python syntax without panicking as well as valid Python syntax. I think currently our parser will never produce an `ast.Raise` node where `exc` is `None` but `cause` is `Some()`, even when analyzing invalid syntax -- but that's an implementation detail of our parser as it is today; we might change the way we do error recovery at any point, so we shouldn't rely on it.

TL;DR: we can't do an early return here; you have to make sure that we also infer the type of `cause` here if it's `Some()`.

File: crates/red_knot_python_semantic/src/types/diagnostic.rs (hunk 6 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/diagnostic.rs:232]: Yeah, I agree that this would be clearer.

File: crates/red_knot_python_semantic/src/types/diagnostic.rs (hunk 5 above)
Comment (carljm) [reply crates/red_knot_python_semantic/src/types/diagnostic.rs:737]: `add_invalid_exception_cause` is correct here; this error is about the "cause", or the `e` in `raise Foo from e`, it's not about catching exceptions.

File: crates/red_knot_python_semantic/src/types/diagnostic.rs (hunk 6 above)
Comment (carljm) [reply crates/red_knot_python_semantic/src/types/diagnostic.rs:232]: Two or three separate codes? On the raising side, there's the raised exception, and there's the cause: the `Foo` and the `e` in `raise Foo from e`. I think if we split it out, we kind of have to go to three, or else one of the two will still have an awkward name: `invalid-exception-raised` doesn't make sense for a bad "cause", because that's not the exception raised.

File: crates/red_knot_python_semantic/src/types.rs (hunk 8 above)
Comment (carljm) [reply crates/red_knot_python_semantic/src/types.rs:1474]: I would like us to increase our use of well-named functions that describe some semantic of a type, rather than just inlining all of it into type inference, particularly when the same logic would be repeated in more than one location. I do think this makes type inference more readable. I think these methods are worth it if they have a sensible/clear name and have 2+ uses. If they have only one use, I agree that's questionable.

As for where they go, I would prefer to keep them on `Type`; I think that's better than a collection of standalone functions which operate on a single `Type`, both in usage ergonomics and in discoverability. I'm not worried about too many methods on `Type`, if we go with the heuristic that methods should generally have 2+ uses, and have a name that clearly represents something semantically meaningful.

Concretely in this case I think those heuristics would suggest keeping `is_exception_class` and `is_exception`, but not `is_raisable` or `can_be_exception_cause`. (I also think `is_none` is fine in general, but as Alex points out, maybe isn't the clearest option here -- though given that `None` is final and that will never change, it really doesn't matter in practice.)

File: crates/red_knot_python_semantic/src/types/diagnostic.rs (hunk 6 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/diagnostic.rs:232]: I think it would be fine to have two codes: `invalid-exception-caught` and `invalid-raise-statement`. One code would be about exception handlers, the other would be about `raise` statements.

Comment (carljm) [reply crates/red_knot_python_semantic/src/types/diagnostic.rs:232]: Sounds reasonable. Maybe just `invalid-raise`? Not sure `-statement` adds value.

File: crates/red_knot_python_semantic/src/types.rs (hunk 8 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:1474]: > Concretely in this case I think those heuristics would suggest keeping `is_exception_class` and `is_exception`, but not `is_raisable` or `can_be_exception_cause`.

`is_exception` is a one-liner, and I honestly find `.is_assignable_to(KnownClass::BaseException.to_instance(db))` more readable, because I don't have to lookup exactly what `is_exception()` is doing under the hood.

`is_exception_class` is a bit more complicated than `is_exception`, because `KnownClass::to_subclass_of()` returns `Option<Type>` rather than `Type`. That's probably a bad signature for `to_subclass_of()` to have, to be honest; I probably made a mistake there when I added it in https://github.com/astral-sh/ruff/commit/ab26d9cf9a5899406e1b4a3555c54e6d0be8ed96#diff-42314c006689490bbdfbeeb973de64046b3e069e3d88f67520aeba375f20e655R1910. Both `KnownClass::to_class_literal()` and `KnownClass::to_instance()` fall back to `Unknown` if they can't find the type being looked up rather than returning an `Option`; for consistency, `KnownClass::to_subclass_of()` should probably return `type[Unknown]` if it can't find the type in question, rather than returning an `Option`.

Comment (carljm) [reply crates/red_knot_python_semantic/src/types.rs:1474]: Ok, I think it's fair to also add the heuristic "encapsulates some non-trivial logic" :) And I agree that making `to_subclass_of` just handle the fallback to `Unknown` internally is a good alternative to `is_exception_class`.

File: crates/red_knot_python_semantic/src/types/diagnostic.rs [hunk 10]
```diff
@@ -233,6 +234,35 @@ declare_lint! {
     }
 }
 
+declare_lint! {
+    /// Checks for `raise` statements that raise non-exceptions or use non-exceptions cause.
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/diagnostic.rs:238]:
```suggestion
    /// Checks for `raise` statements that raise non-exceptions or use invalid
    /// causes for their raised exceptions.
```

File: crates/red_knot_python_semantic/src/types/diagnostic.rs [hunk 11]
```diff
@@ -233,6 +234,35 @@ declare_lint! {
...
+    ///
+    /// ## Why is this bad?
+    /// Only classes inherit from and instances of `BaseException` can be raised.
+    /// Only classes inherit from and instances of `BaseException` and `None`
+    /// can be used as exception cause.
+    /// Attempts to violate this restriction will result in a `TypeError` at runtime.
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/diagnostic.rs:244]:
```suggestion
    /// Only subclasses or instances of `BaseException` can be raised.
    /// For an exception's cause, the same rules apply, except that `None` is also
    /// permitted. Violating these rules results in a `TypeError` at runtime.
```

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 12]
```diff
@@ -2194,8 +2188,34 @@ impl<'db> TypeInferenceBuilder<'db> {
...
+        let can_be_exception_cause = UnionType::from_elements(self.db, [can_be_raised, none]);
+
+        if let Some(raised) = exc {
+            let raised_type = self.infer_expression(raised);
+
+            if !raised_type.is_assignable_to(self.db, can_be_raised) {
+                self.diagnostics
+                    .add_invalid_exception_raised(self.db, raised, raised_type);
+            }
+        }
```
Comment (MichaReiser) [inline crates/red_knot_python_semantic/src/types/infer.rs:2209]:
```suggestion

        let can_be_exception_cause = UnionType::from_elements(self.db, [can_be_raised, none]);

        if let Some(raised) = exc {
            let raised_type = self.infer_expression(raised);

		        let can_be_raised =
	            UnionType::from_elements(self.db, [base_exception_type, base_exception_instance]);
            if !raised_type.is_assignable_to(self.db, can_be_raised) {
                self.diagnostics
                    .add_invalid_exception_raised(self.db, raised, raised_type);
            }
        }
```

File: crates/red_knot_python_semantic/src/types/diagnostic.rs [hunk 13]
```diff
@@ -233,6 +234,35 @@ declare_lint! {
...
+    ///
+    /// ## Example
+    /// ```python
+    /// raise int from str()
+    /// ```
+    ///
+    /// Use instead:
+    /// ```python
+    /// raise ZeroDivisionError() from RuntimeError
+    /// ```
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/diagnostic.rs:254]:
```suggestion
    /// ## Examples
    /// ```python
    /// def f():
    ///     try:
    ///         something()
    ///     except NameError:
    ///         raise "oops!" from f
    ///
    /// def g():
    ///     raise NotImplemented from 42
    /// ```
    ///
    /// Use instead:
    /// ```python
    /// def f():
    ///     try:
    ///         something()
    ///     except NameError as e:
    ///         raise RuntimeError("oops!") from e
    ///
    /// def g():
    ///     raise NotImplementedError from None
    /// ```
```

File: crates/red_knot_python_semantic/src/types.rs [hunk 14]
```diff
@@ -1443,6 +1443,14 @@ impl<'db> Type<'db> {
         }
     }
 
+    pub(crate) fn is_exception_class(self, db: &'db dyn Db) -> bool {
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:1446]: This method also returns `true` (and is correct to do so) for `Type::Any`, `Type::Unknown` and `Type::Todo`, so if we're going to keep it for now, it would be better to give it a more precise name than `is_exception_class`:

```suggestion
    pub(crate) fn is_assignable_to_subclass_of_base_exception(self, db: &'db dyn Db) -> bool {
```

But I would still favour just simplifying the signature of `KnownClass::to_subclass_of()` so that we don't need this method at all

File: crates/red_knot_python_semantic/src/types.rs [hunk 15]
```diff
@@ -2315,10 +2315,11 @@ impl<'db> KnownClass {
             .unwrap_or(Type::Unknown)
     }
 
-    pub fn to_subclass_of(self, db: &'db dyn Db) -> Option<Type<'db>> {
+    pub fn to_subclass_of(self, db: &'db dyn Db) -> Type<'db> {
         self.to_class_literal(db)
             .into_class_literal()
             .map(|ClassLiteralType { class }| Type::subclass_of(class))
+            .unwrap_or(Type::Unknown)
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:2322]:
```suggestion
            .unwrap_or(Type::subclass_of_base(ClassBase::Unknown))
```

**Review by AlexWaygood — APPROVED** [review]
Nice, thank you!

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 16]
```diff
@@ -2198,8 +2195,30 @@ impl<'db> TypeInferenceBuilder<'db> {
...
-        self.infer_optional_expression(exc.as_deref());
-        self.infer_optional_expression(cause.as_deref());
+
+        let base_exception_type = KnownClass::BaseException.to_subclass_of(self.db());
+        let base_exception_instance = base_exception_type.to_instance(self.db());
+        let none = KnownClass::NoneType.to_instance(self.db());
+
+        let can_be_raised =
+            UnionType::from_elements(self.db(), [base_exception_type, base_exception_instance]);
+        let can_be_exception_cause = UnionType::from_elements(self.db(), [can_be_raised, none]);
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:2205]:
```suggestion

        let can_be_raised =
            UnionType::from_elements(self.db(), [base_exception_type, base_exception_instance]);
        let can_be_exception_cause = UnionType::from_elements(self.db(), [can_be_raised, Type::none(self.db())]);
```
